### PR TITLE
Incentives: Implements "Mining" - diverting a portion of fees to proposers

### PR DESCRIPTION
**Pull request by jannotti**
This is only a draft, in the sense of code that isn't ready for merge AND **an idea that isn't ready for merge**. There is no release timeline for it.

I’m using the word “mining” to mean the portion of incentives for proposing a block that comes from fees. Today’s fees are tiny, but we should introduce this mechanism now, to show how incentives will need to work eventually. The mining mechanism must be built into the L1 protocol, because it diverts funds from the FeeSink. The following should be sufficient to support paying out these incentives, but it lacks the absenteeism protections that are needed because of "Bonus" payouts - the extra incentives that will be paid by the Foundation that do not come from fees.

- [x] If proto.EnableMining, add the current Proposer and the FeesCollected to the block header.
- [x] Choose proto.MiningPercent which is the percentage of algos paid in fees that should go to the proposer of the block from the FeeSink.
- [x] At the start of block n+1, move MiningPercent * block[n].FeesCollected from the FeeSink to block[n].Proposer.
- [x] Outlaw any other payments from FeeSink to ensure the funds to be moved must be present.

The movement is done at the start of block n+1 instead of during block n because a multi-participant node builds a block before it knows which account might actually be the proposer. The actual proposer is determined later, in the agreement code. It is possible we could perform the movement at the end of block n instead, by going back into the block and modifying the state delta of the ValidatedBlock when agreement calls WithSeed.  However, in the current code, agreement does not have access to the ValidatedBlock‘s state delta.  The choice of n+1 seems simpler, and funds are never invisible in this design. They sit in the FeeSink “between” rounds.

### CODE REVIEW

**Comment by joe-p** [comment]
One thing that should be considered is if and how the protocol determines if the proposer is acting honestly. By simply moving `MiningPercent * block[n].FeesCollected` to the proposer, it incentives people to create 0-transaction blocks. My understanding is that the committees do not check whether the block has transactions or not, but rather if the transaction it does have are valid or not. Of course, right now there probably isn't much cost to be saved by doing so for an individual, but it it could be problematic at scale. For example, if a delegation service proposes empty blocks, they might have a x% reduction in the cost of running their nodes, meaning they can offer the delegators more of a return. x% becomes larger and larger as the network scales and validating blocks involve more computation. An end-user that doesn't care about the network and just wants to earn more ALGO will be more incentivized to go with this 0-txn delegation services, rather than a legitimate one that offers less in rewards.

**Comment by jannotti** [comment]
> By simply moving `MiningPercent * block[n].FeesCollected` to the proposer, it incentives people to create 0-transaction blocks.

If you create a 0 txn blocks, the block's `FeesCollected` will be zero. You will make no money.

**Comment by joe-p** [comment]
> If you create a 0 txn block, the block's FeesCollected will be zero. You will make no money.

🤦 Yes I was still thinking in the mindset of non-fee-based incentives.

**Comment by pao-beep** [comment]
Doesn't this centralized block production, the wealthy proposers get even richer.

**Comment by MochaN3rd** [comment]
> Doesn't this centralize block production, the wealthy proposers get even richer.

What if the fee was split between the voters and the block proposers? I run a node and vote on significantly more blocks than I propose

**Comment by jannotti** [comment]
> Doesn't this centralize block production, the wealthy proposers get even richer.

Does it centralize block production? I don't see how. If this offers meaningful returns, it helps any potential participant defray the costs of participating. Today, these costs are born by the proposer, so surely the current policy leads to more centralization, as only the rich would be willing to pay those costs for fun.

Will wealthy proposers get even richer? At the current rates, this will only defray infrastructure costs. They will get poor more slowly.  But jokes aside, it seems proper that the people running nodes are compensated by the money being paid for the work being done by those nodes.

**Comment by pao-beep** [comment]
> > Doesn't this centralize block production, the wealthy proposers get even richer.
> 
> Does it centralize block production? I don't see how. If this offers meaningful returns, it helps any potential participant defray the costs of participating. Today, these costs are born by the proposer, so surely the current policy leads to more centralization, as only the rich would be willing to pay those costs for fun.
> 
> Will wealthy proposers get even richer? At the current rates, this will only defray infrastructure costs. They will get poor more slowly.  But jokes aside, it seems proper that the people running nodes are compensated by the money being paid for the work being done by those nodes.

Okay this doesn't make any distinction between amounts staked does it. Meaning if I stake 1k algos and produce a big block with more fees and say another proposer stakes 1million algos and produces a small block with small fees it's fair game. I get to keep the fees regardless of my stake?

**Comment by jannotti** [comment]
> > Doesn't this centralize block production, the wealthy proposers get even richer.
> 
> What if the fee was split between the voters and the block proposers? I run a node and vote on significantly more blocks than I propose

In the end, it will all come out the same.  When you participate, you represent the same fraction of voters as you represent among proposers.  You do vote much more often than you propose, since the cert committee has something like 1500 voters.  But the incentive would have to be split among those 1500!  So over enough blocks, your payout will be the same.

This payout scheme is a bit "lumpier", but there are technical challenges to the scheme you propose that make me very willing to accept the lumpiness. In particular, the network reaches consensus about _blocks_, but not about the _voters_ on each block.  Two different nodes might see a different set of votes --- they just each have to see enough.  To perform payouts to voters, we would need to consume block space to write down and reach consensus about the voters.

**Comment by jannotti** [comment]
> Okay this doesn't make any distinction between amounts staked does it. Meaning if I stake 1k algos and produce a big block with more fees and say another proposer stakes 1million algos and produces a small block with small fees it's fair game. I get to keep the fees regardless of my stake?

Correct. By having less stake you will propose less often (just as today), but when you do, you get all of the incentive paid here.

**Comment by cswenor** [comment]
This is great. A couple things that would be nice is the ability to set a configuration at what portion of the fees get given to the block producer and how much go to a treasury. Also it would be nice to allow for nodes to opt out of the reward so that the foundation could run nodes without taking the rewards.

**Comment by jannotti** [comment]
> This is great. A couple things that would be nice is the ability to set a configuration at what portion of the fees get given to the block producer and how much go to a treasury.

The consensus parameter `MiningPercent` decides how much of the fee goes to the proposer.

> Also it would be nice to allow for nodes to opt out of the reward so that the foundation could run nodes without taking the rewards.

That's a good idea. Perhaps a proposer can write any address they'd like into the block's `Proposer` field. They could write the FeeSink's address to leave the funs alone.

**Comment by pao-beep** [comment]
> > Okay this doesn't make any distinction between amounts staked does it. Meaning if I stake 1k algos and produce a big block with more fees and say another proposer stakes 1million algos and produces a small block with small fees it's fair game. I get to keep the fees regardless of my stake?
> 
> Correct. By having less stake you will propose less often (just as today), but when you do, you get all of the incentive paid here.
> 

Nice! I like it

**Comment by ctibo** [comment]
> > Okay this doesn't make any distinction between amounts staked does it. Meaning if I stake 1k algos and produce a big block with more fees and say another proposer stakes 1million algos and produces a small block with small fees it's fair game. I get to keep the fees regardless of my stake?
> 
> Correct. By having less stake you will propose less often (just as today), but when you do, you get all of the incentive paid here.

Does this mean smaller participants would benefit of pooling their stake together to propose more blocks? 

They could do it using a smart contract, to make sure their stake is safe. The pool contract would still participate from a single machine though. So we'd eventually centralize consensus.

It seems we're trying to get rid of relays to better decentralize the protocol, but we'll incentivize people to create the biggest pools and win the most blocks.

I might be overthinking it...

**Comment by jannotti** [comment]
> > > Okay this doesn't make any distinction between amounts staked does it. Meaning if I stake 1k algos and produce a big block with more fees and say another proposer stakes 1million algos and produces a small block with small fees it's fair game. I get to keep the fees regardless of my stake?
> > 
> > 
> > Correct. By having less stake you will propose less often (just as today), but when you do, you get all of the incentive paid here.
> 
> Does this mean smaller participants would benefit of pooling their stake together to propose more blocks?
> 
> They could do it using a smart contract, to make sure their stake is safe. The pool contract would still participate from a single machine though. So we'd eventually centralize consensus.
> 
> It seems we're trying to get rid of relays to better decentralize the protocol, but we'll incentivize people to create the biggest pools and win the most blocks.
> 
> I might be overthinking it...

Two people pooling their algos will get the same total rewards as if they operated separately. So in that sense, there's no benefit.

On the other hand, there would be some cost savings from running fewer nodes.

I can see some benefit to pooling, to save costs and perhaps have the nodes that do run be run more professionally. I can also see downsides, as large nodes would be a bigger impact to consensus if they go down.

**Comment by HashMapsData2Value** [comment]
I'd like to confirm if my own understanding of consensus is correct.

Given the high performant nature of the PPoS protocol the cost of running a node does not scale with stake, right?  E.g., Bob has 10k Algo and Eve has 1m Algo, but they can both go and buy the same computer to run an Algorand node on it. They will suffer the same costs.

But since Eve has 1m Algo she will propose blocks more often than Bob. If we incentivize node running by rewarding block proposals, Eve will be rewarded more often. Since Algo is a limited resource, over time, Eve will accumulate more of an ultimately finite resource, compounding and potentially snowballing over time.

On the other hand, there is another equally important and

**Comment by HashMapsData2Value** [comment]
Given the high performant nature of the PPoS protocol the cost of running a node does not scale with stake, right? It's a "single coin toss" to get the result of the VRF? And since we have a fixed max opcode budget, we have a minimum level of compute hardware the average node needs to maintain, but there's no point in running a super strong computer since everyone else will be too slow? 

E.g., if Bob has 10k Algo and Eve has 1m Algo, they can both go and buy the same computer to run an Algorand node on it. They will suffer the same costs But since Eve has 1m Algo she will propose blocks more often than Bob. If we incentivize node running by rewarding block proposals, Eve will be rewarded more often. Since Algo is a limited resource, over time, Eve will accumulate more of an ultimately finite resource, compounding and potentially snowballing over time.

I'm all for rewarding people for their contribution to maintaining the integrity of Algorand, but it'll definitely be tricky setting the %. The higher it is the more we incentivize node runners, but the faster we could get to a "snowballing" scenario. And if it's too low we might not be able to smaller holders to go through the effort of running a node. (And while individually a smaller holder might not be impactful, as a large group they would be essential in decentralizing the network.)

On the other hand, there is another resource that is incredibly important to Algorand and whose cost does scale with usage - networking. Since PPoS is so efficient and we have a fixed opcode budget setting the standard on compute, the bottle neck lies in networking. The more you contribute positively (i.e., helping to connect to many peers, not acting like a bottleneck, etc) the higher the performance of Algorand, but also the higher your cost also  _ought_ to be. 

Since we are switching over to P2P gossip structure anyway, everyone will be contributing to Algorand's networking and be able to get funds. But the limited resource providing sybil-resistance should not be Algo, but rather networking. The big Algo holders are already getting some incentives to engage in consensus with their nodes and protect their bag.

I dont have any good suggestions here for how to create an incentive structure for networking. A pitfall would be rewarding node runners in developing countries at the risk of punishing those in networking-challenged ones. We'd need to have a way to take into consideration that somehow.

Finally, another thing to consider is archive nodes. Maintaining full or partial archive of Algorand might also be a way to get rewarded from the fees. Storage is another resource where the more you contribute, the more you get rewarded, but the higher your costs will be as well.

TL;DR: The cost of consensus does not scale with Algo bag. Perhaps we should reward in a different way, like networking or storage.

**Comment by pbennett** [comment]
> That's a good idea. Perhaps a proposer can write any address they'd like into the block's `Proposer` field. They could write the FeeSink's address to leave the funds alone.

Makes sense for payout but if it replaced the on-chain proposer, this would be for unfortunate for the things that track the proposer of each block for consensus participation metrics.

**Comment by pao-beep** [comment]
@HashMapsData2Value the network incentivization is going to be in place for relay nodes. This rewarding for participating nodes is merely to get more people involved. It could be that a lower stake node gets more fees if during the time it was chosen there was a bigger block

**Comment by pkariz** [comment]
>  Perhaps a proposer can write any address they'd like into the block's Proposer field

I don't know the details of how things currently work, but i don't think the proposer should have the power to send me the fees (eg. evil proposer wants to get me in trouble with regulations - it's true that transactions are non-interactive so anyone can send coins to me at any time, but maybe "mining" coins will be looked at differently). To counter that maybe just "donate fees to feesink" option should be allowed (like a binary config option).

>  I can see some benefit to pooling, to save costs and perhaps have the nodes that do run be run more professionally

One other benefit is also more frequent and therefore more consistent payouts, especially for smaller participants.

>  Perhaps we should reward in a different way, like networking or storage.

My guess is that it would be hard to evaluate and prove how much someone is contributing to networking or storage.

**Comment by jannotti** [comment]
> > That's a good idea. Perhaps a proposer can write any address they'd like into the block's `Proposer` field. They could write the FeeSink's address to leave the funds alone.
> 
> Makes sense for payout but if it replaced the on-chain proposer, this would be for unfortunate for the things that track the proposer of each block for consensus participation metrics.

Currently, the proposer is not in the block, per se. All nodes _do_ agree on the proposer (they must, so they can check the proposed BlockSeed is correct), but they have that information "on the side" in the messages they collect while confirming a block.  This PR adds an explicit field, called `Proposer` to the BlockHeader, and _that_ is the address that gets paid. If we wanted the proposer to be able to decline the incentive, it would be that field that they'd change. The network would still know who the real proposer was.

If this is deemed desirable, I suppose I'd want to rename the field `Miner` or `MinerRewards`.  And, at the protocol level we're talking about the names of msgpack fields. So it would be `mnr` instead of `prp`, or something like that.

**Comment by jannotti** [comment]
> > Perhaps a proposer can write any address they'd like into the block's Proposer field
> 
> I don't know the details of how things currently work, but i don't think the proposer should have the power to send me the fees (eg. evil proposer wants to get me in trouble with regulations - it's true that transactions are non-interactive so anyone can send coins to me at any time, but maybe "mining" coins will be looked at differently). To counter that maybe just "donate fees to feesink" option should be allowed (like a binary config option).

Since the "real" proposer is always going to be known (see my last comment), I don't see how this could possibly be interpreted any differently that a normal `pay` transaction. Someone _else_ wrote your address down and so you got algos.

> > Perhaps we should reward in a different way, like networking or storage.
> 
> My guess is that it would be hard to evaluate and prove how much someone is contributing to networking or storage.

That's exactly right. The _protocol_ can't reward anything that's invisible to the protocol. Networking and storage (beyond that needed to evaluate transactions) are both invisible. And the storage that _is_ needed to evaluate transactions is the same size for everyone.  An off-chain system could perhaps monitor some of these things and hand out money, but I don't think it could be made part of the protocol without massive changes (things like Filecoin's Proof of Storage).

**Comment by HashMapsData2Value** [comment]
> @HashMapsData2Value the network incentivization is going to be in place for relay nodes. This rewarding for participating nodes is merely to get more people involved. It could be that a lower stake node gets more fees if during the time it was chosen there was a bigger block

I'm not sure I understand this argument since this should average out across many blocks over time, and those with more Algo would simply propose more blocks? 

> That's exactly right. The protocol can't reward anything that's invisible to the protocol. Networking and storage (beyond that needed to evaluate transactions) are both invisible. And the storage that is needed to evaluate transactions is the same size for everyone. An off-chain system could perhaps monitor some of these things and hand out money, but I don't think it could be made part of the protocol without massive changes (things like Filecoin's Proof of Storage).

What about inflation? If we could inflate the Algo supply in such a way that the stake proportions/distributions between participation nodes are maintained (and preventing the snowballing of a whale vs other node runners), it would incentivize running a node to avoid your holdings losing value to inflation.

Off the top of my head the problems I see are:
- 1) dealing with fixed minimum balance requirements  in an inflationary environment
- 2) confirming that a node is non-maliciously engaged in consensus (i.e. their account is activated but they're not actually helping out).

**Comment by joe-p** [comment]
One thing that I think should be mentioned is this fundamentally changes how fee under congestion will play out in the network. Rational miners will treat fees as a first-price auction which means users will not have a known dominant strategy and will need to outbid each other. I think it's worth considering a mechanism similar to EIP1559, where `MiningPercent` decreases as the network gets congested, thus more of the fees are sent to the `FeeSink` under congestion.

**Comment by jannotti** [comment]
> One thing that I think should be mentioned is this fundamentally changes how fee under congestion will play out in the network. Rational miners will treat fees as a first-price auction...

Not just under congestion.

> I think it's worth considering a mechanism similar to EIP1559, where `MiningPercent` decreases as the network gets congested, thus more of the fees are sent to the `FeeSink` under congestion.

Since your concern is true at all times, I don't see the value in doing that. It would also be very difficult. The level of congestion is not something that the protocol reaches consensus on. So it works be hard to validate the choice of `MiningPercent` if it changes with congestion.

**Comment by pao-beep** [comment]
@HashMapsData2Value yes but not the fees, that's unpredictable. The business of the network is unpredictable. Over a long enough period yeah but it's not straightforward. Besides like jannoti said It's meant to defray cost but if there is enough fees to be collected you might end up making just a little bit more. Looking forward to knowing how that percentage to proposers is determined to gauge the " just enough to defray cost" and not introduce some zero sum profit games

**Comment by pkariz** [comment]
>  Someone else wrote your address down and so you got algos

Miners could stay anonymous, put their other non-mining account as the receiver and pretend it's not them. To avoid such tricks the tax company would need to tax those algos as mining algos imo, so i guess it could create some problems for the receiver.

**Comment by pbennett** [comment]
> That's exactly right. The _protocol_ can't reward anything that's invisible to the protocol. Networking and storage (beyond that needed to evaluate transactions) are both invisible.

Since we're talking about big protocol changes here - might as well spitball a bit?  In terms of 'is a node an archival node' - isn't that required and visible as part of providing 'catchup/old blocks'?  Just spitballing, but I wonder if some mechanism could be defined where a node needing to catchup could ask for a more formal (and efficient) packaging of blocks w/ proof - allow multiple proposers (w/ random verifiable winner) and there be some payment mechanism already set up.
Perhaps an escrow prior to catchup, or perhaps as part of part. key online registration having submitted a delegated signature against a 'pay for catchup lsig' that's tied into a global payout contract preventing abuse/double-charges.

**Comment by joe-p** [comment]
> Not just under congestion.

How so? Without scarce block space, why would anyone pay more than `MinTxnFee`? Miners are incentivized to include as many transactions as possible, regardless of the fee.

> The level of congestion is not something that the protocol reaches consensus on. So it works be hard to validate the choice of MiningPercent if it changes with congestion.

I'm not suggesting the `MiningPercent` is calculated based on mempool congestion, but based on the historical block sizes. For example, every node keeps track of how many full blocks there have been recently and adjusts `MiningPercent` accordingly.

**Comment by jannotti** [comment]
> > Not just under congestion.
> 
> How so? Without scarce block space, why would anyone pay more than `MinTxnFee`? Miners are incentivized to include as many transactions as possible, regardless of the fee.

If you believe that nodes have modified algod to choose the highest paying transactions first in times of congestion, you probably believe that it works by maintaining a priority queue using fee/byte as the "priority". (I wouldn't expect them to have two different strategies to build blocks - that would imply that just as a block gets full, they'd have to throw it all away and rebuild it while taking fees into account.)

If that's the case, you can get to the front of the queue by paying more.

**Comment by joe-p** [comment]
> If that's the case, you can get to the front of the queue by paying more.

Sure, but we're talking about two different things then. When there is free block space, the dominant strategy to get *included* in a block is to simply pay `MinTxnFee`. 

I suppose if we really wanted to optimize the strategy for the users wishing to get included in the block we'd need a separate base fee and priority fee just like EIP-1559. 

We have the advantage of implementing incentives after ETH has gone through EIP-1559, so why not learn from them and implement an EIP-1559-like mechanism (I say like because we don't have block rewards, so the non-congested fee mechanism will need to be different) right from the start?

File: ledger/eval/eval.go [hunk 1]
```diff
@@ -1160,12 +1167,27 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, evalParams *
...
 }
 
+func (cs *roundCowState) takeFee(tx *transactions.Transaction, senderRewards *basics.MicroAlgos, ep *logic.EvalParams) error {
+	err := cs.Move(tx.Sender, ep.Specials.FeeSink, tx.Fee, senderRewards, nil)
+	if err != nil {
+		return err
+	}
+	// transactions from FeeSink should be exceedingly rare. But we can't count
+	// them in feesCollected because there are no net algos added to the Sink
+	if tx.Sender == ep.Specials.FeeSink {
```
Comment (bbroder-uji) [inline ledger/eval/eval.go:1177]: nonetheless, if I'm proposing blocks, is it my fault that I only have a bunch of FeeSink transactions?  In that situation the proposer gest nothing right?

**Comment by joe-p** [comment]
One other thing that needs to be considered is that miners will benefit from congestion, thus they might be incentivized to cause it. Let's say it costs $0.10 to fill a block. If a coalition of miners has 1% of online stake, they are incentivized to fill the blocks as long they believe they can generate at least $10 of fee revenue in their proposed blocks. Of course, this largely depends on how much users are actually willing to pay to get their transactions in a block. 

As such I would propose we consider increasing `MinTxnFee` or having the fee cost scale with transactions size (much like current congestion fees).

File: ledger/eval/eval.go (hunk 1 above)
Comment (jannotti) [reply ledger/eval/eval.go:1177]: It's possible we could go even further than this PR currently proposes and ban _all_ FeeSink transactions. (We only ban pays now.). I don't see any real problems doing that, actually.

Comment (bbroder-uji) [reply ledger/eval/eval.go:1177]: I think that makes more sense than marking them as paying 0 mining rewards.

**Comment by jannotti** [comment]
> One other thing that needs to be considered is that miners will benefit from congestion, thus they might be incentivized to cause it. Let's say it costs $0.10 to fill a block. If a coalition of miners has 1% of online stake, they are incentivized to fill the blocks as long they believe they can generate at least $10 of fee revenue in their proposed blocks. Of course, this largely depends on how much users are actually willing to pay to get their transactions in a block.

There's a difficult interplay here between what is and is not determined by consensus (congestion fees are not determined by consensus). Putting that aside, this is a pretty bold strategy. It fails miserably if people aren't willing to pay many multiples of the `MinTxnFee`. And when it fails, you just lose money, before you even know if it will work.
 
> As such I would propose we consider increasing `MinTxnFee` or having the fee cost scale with transactions size (much like current congestion fees).

I wish `MinTxnFee` depended on length, and `MinTxnFee` certainly needs to grow someday in order to be the sole source of revenue to sustain the blockchain. But I suspect it would wreak havoc on existing code if `global MinTxnFee` was not a constant. For one thing, it would no longer be legal to call that in a LogicSig. - and the vast majority of LogicSigs should and do call it.

**Comment by pbennett** [comment]
> One other thing that needs to be considered is that miners will benefit from congestion, thus they might be incentivized to cause it. Let's say it costs $0.10 to fill a block. 

The fee escalation is at the node level though (which they could trivially game particularly if you're talking to *their* nodes), but to truly raise the fees for the entire network they'd basically have to legitimately spam the network with legitimate transactions so they pay back their spam txn fee but also only IF they win the proposed block.  If spamming the network though, THEY have to pay the increased fees as well - and at least for now, I'd guess a LOT of applications won't work too well with escalating fees (or fail outright).  It just seems a big gamble since it also disincentivizes normal network use - which they want to profit from as well.

**Comment by joe-p** [comment]
> There's a difficult interplay here between what is and is not determined by consensus (congestion fees are not determined by consensus).

Right sorry, by congestion I just mean "making block space scarce". If at any given time a user doesn't know if the next block will be full or not, they are incentivized to treat their fee as a first-price auction bid. Otherwise they risk not getting included. 

>  Putting that aside, this is a pretty bold strategy. It fails miserably if people aren't willing to pay many multiples of the MinTxnFee. And when it fails, you just lose money, before you even know if it will work.

I agree and I think its hard to properly model. To me the main problem is that it's still incredibly cheap to do this. A miner coalition can fill many blocks for a relatively low amount of money, which gives time for users to adjust their bidding strategy. 

>  but also only IF they win the proposed block.

Right, but they roughly know how often they will win the proposed block. If the coalition has 1% of the online stake then they know they need to pay the fees for roughly 100 blocks before they can expect to see a profit. As long as their profit is larger than the cost to fill those 100 blocks, they don't care that they need to pay all those fees. 

> If spamming the network though, THEY have to pay the increased fees as well 

Yes, but they can simply spam the network to keep blocks full but not let the number of pending blocks exceed 1. Still extra fees, but not much. They can also use inner transactions for a majority of the spam, which are hard-coded to be 0.001 ALGO even under mempool congestion. 

>  It just seems a big gamble since it also disincentivizes normal network use - which they want to profit from as well.

The risk/reward is heavily weighted towards reward though. Risk losing out on a fractions of a cent (and they operate nodes at a loss) to potentially start turning a profit.

**Comment by GhostOfMcAfee** [comment]
> Also it would be nice to allow for nodes to opt out of the reward so that the foundation could run nodes without taking the rewards.

Is that necessary? If AF gets X Algo (from fee sink) as payment in round n+1 but X Algo also goes to the fee sink in round n, what is your suggestion accomplishes besides assuring the fee sink increases. My preference would be for a simpler solution unless there is a very good reason for adding complexity.

**Comment by GhostOfMcAfee** [comment]
While I do appreciate the effort to make this a protocol level implementation, part of me wonders if there it is not better to make this a function of Governance. Specifically, Governors that participate in consensus could get a boost in governance rewards based on their number of proposals and committee votes. 

First, it would get around the issue of AF participating (since AF does not participate in governance). Second, it would simplify things by not requiring a protocol change. Third, at the current time, I don't think protocol level incentives based on fees are all that attractive. Hence, I don't think they are likely to do much to bring an influx of nodes in the short term. On the other hand, I think our incentives for Governance are out of wack and give far too much to people who do too little, and thus we incentivize unproductive behavior. At a basic level, I realize why you cannot long term incentivize a person with more than the block fees. However, I also don't know why people who largely check out for 2 1/2 months every quarter and then swoop in to invest minimal time to vote should get such a large amount of rewards either.

**Comment by cswenor** [comment]
@GhostOfMcAfee I think you have strong points, but I don't believe that has to do with the code being able to do this functionality. @jannotti addressed in his first message that this wouldn't make a huge impact in the near term, but it is something that would be beneficial if it was a setting in the code.
The points you bring up should continue to be debated though, I just don't think this is the right venue for it.

**Comment by pao-beep** [comment]
> One thing that I think should be mentioned is this fundamentally changes how fee under congestion will play out in the network. Rational miners will treat fees as a first-price auction which means users will not have a known dominant strategy and will need to outbid each other. I think it's worth considering a mechanism similar to EIP1559, where `MiningPercent` decreases as the network gets congested, thus more of the fees are sent to the `FeeSink` under congestion.

I think there'll be perverse incentives here with this introduction of mining fee i.e MEV for fees, altering transaction order for most fees and working with searchers to make that happen. I don't think we should import that dumpster from Ethereum. 

Maybe a flat reward to proposers should suffice regardless of fees in the block. The flat reward can be adjusted every 3 months through governance depending how much people are participating in consensus. If less people are participating increase the reward. Governors can decide whether it makes sense to adjust given that the amount of fees in the fee sink can be scarce

**Comment by pao-beep** [comment]
@jannotti in regards to the ability to reorder txns for monetary gain being already present, not moving with the proposal of percentage of fee rewards based on fees in block at least prevents such mev games on fees.

Someone can be spammed and  get more fees through fee per byte if they run a participating node as a service. I think we should try as much to prevent dPos from creeping in. Maybe I'm overthinking it 

I still think a flat fee reward from fee sink to a proposer is good enough to prevent some of these types of issues and periodic adjustment (320 blocks?) Can be done on the reward whether through governance or automatically through code. I think this might I think it'll be easier through code. A formula can be devised that increases flat fee rewards when total stake is decreasing.

**Comment by joe-p** [comment]
Now that I think about it a miner coalition seeking profit wouldn't need to congest the network (in fact they shouldn't). In that scenario rational users would simply just bid `MinTxnFee` + 1 uALGO to get included in the block. Instead, a miner coalition could simply agree to not add transactions under a certain threshold. With this approach, I think the coalition would need closer to say 10% of the stake to have a meaningful impact on user bid strategy. It basically comes down to how much the user is willing to pay for 100% guarantee of block inclusion vs `100 - coalitionStake`% chance of block inclusion. For economic actors, such as arb bots, I'd assume that the amount they are willing to pay is non-zero at some threshold, but it's unclear what that is.

Honestly I think the main problem is that implementing this as-is now offers no benefits (I'd hardly call making 1/10th of a cent per block a benefit) but puts user-experience at risk due to the fundamental change in miner incentives. 

As soon as participating becomes profitable (with some reasonable amount of ALGO) then this is no longer a concern. Any rational economic actor will simply delegate rather than stand up their own node to join a profit coalition. 

Some napkin math:

I think it's reasonable to say that you need 1% of online stake to turn a profit, so I'll use that number. With 1% of stake you would currently propose around 250 blocks per month. Running a node that meets minimum requirements with a cloud provider is roughly $100/mo. This means blocks should generate roughly $4 in revenue for a miner. This means, given the current network activity, `MinTxnFee` would need to have a 266x increase. This is assuming 100% of fees go to the miner, which obviously they shouldn't. Of course this number improves as you increase what percentage of stake you think should be profitable, but doing so inherently increases the centralization risk.

**Comment by jannotti** [comment]
> Instead, a miner coalition could simply agree to not add transactions under a certain threshold. 

That can be done today, by having an extra protocol txn sharing network, like MEV-Boost.  This proposal does not change that.

> As soon as participating becomes profitable (with some reasonable amount of ALGO) then this is no longer a concern. Any rational economic actor will simply delegate rather than stand up their own node to join a profit coalition.


> I think it's reasonable to say that you need 1% of online stake to turn a profit, so I'll use that number. With 1% of stake you would currently propose around 250 blocks per month.

Your napkin is broken. 😄 

In a month, we propose about 750k blocks (30*24*60*60/3.5). So with 1% of stake, you propose about 7,500 times.

**Comment by joe-p** [comment]
> That can be done today, by having an extra protocol txn sharing network, like MEV-Boost. This proposal does not change that. (demand a separate payment to be included)

Yes of course, but implementing it in the protocol makes it significantly easier. 

> Your napkin is broken. 😄
>
> In a month, we propose about 750k blocks (30 * 24 * 60 * 60 / 3.5). So with 1% of stake, you propose about 7,500 times.

Lol yes sorry, used the calculation for a day. Still, would require a 10x to `MinTxnFee` to allow 1% stake to turn a profit (again, assuming 100% fees go to miners which they won't). You've mostly talked me off my ledge as per usual, but I still feel a bit uneasy about implementing this without increasing `MinTxnFee` alongside it. Miner incentives aside, it also incentives stake to centralized. It's nice that delegation is non-custodial, but given the state of the ecosystem and the complexity of running an Algorand delegation service that scales, I'm concerned it will be too easy for one entity to become the only delegation service.

**Comment by d13co** [comment]
Made a spreadsheet with txn fees per proposer for a period I was monitoring for some friends: [Google Sheets - Proposer fees 30685000-32157115](https://docs.google.com/spreadsheets/d/1EXXdO-9BZ0VVodmOmdBsCAUCzvU4bn2gXctVIjnduIw/edit#gid=1927799934).

Current_balance is a snapshot just a while ago, will be off for some addresses.

The proj_apr is a percentage already... Would be good to consider having an opt-out of rewards for proposers like the foundation. Maybe a rolling proposer fee pool where ineligible proposers leave the fees for the first eligible proposer to take.

For context, the node I am participating with has total stake in the ~800m-900m ALGO range and the total fees corresponding to me over ~58 days were in the ~20-24 ALGO range. The node cost was $54 for 60 days.

**Comment by ismael-elatifi** [comment]
> * MiningPercent * block[n].FeesCollected

This incentives putting as many transactions as possible in the proposed block.   
So would this penalize "big" transactions in favor of "small" transactions?

**Comment by joe-p** [comment]
> This incentivizes putting as many transactions as possible in the proposed block (which seems desirable at first).
But would this penalize "big" transactions in favor of "small" transactions (assuming we could put more small txs than big ones in a block) ?

This would only be a concern when black space is scarce (ie. congested network). Once that happens the miners are incentivized to take whatever transactions make them the most money thus they will run a first-price auction to be included in the block. Once this occurs, bigger transactions will likely have to bid higher for the same probability of inclusion as smaller transactions.

This is only a concern once congested hits, so we're pretty far away from that. If the network does get congested consistently at some point in time, adopting EIP-1559 or whatever the latest fee mechanism is at the time.

**Comment by pao-beep** [comment]
How do we prevent proposer n +1 from refusing to send the mining fee to proposer n

**Comment by jannotti** [comment]
It is carried out automatically when the next block occurs. The next
proposer doesn't explicitly do anything.

**Comment by jannotti** [comment]
> > This is great. A couple things that would be nice is the ability to set a configuration at what portion of the fees get given to the block producer and how much go to a treasury.
> 
> The consensus parameter `MiningPercent` decides how much of the fee goes to the proposer.
> 
> > Also it would be nice to allow for nodes to opt out of the reward so that the foundation could run nodes without taking the rewards.
> 
> That's a good idea. Perhaps a proposer can write any address they'd like into the block's `Proposer` field. They could write the FeeSink's address to leave the funds alone.

To follow this up - There are going to be a couple rules that participants will have to follow in order to receive incentives.  There will be a minimum account size, maximum account size, and a little extra fee that will be paid when going online from an offline state (whether it's your first keyreg, or because you were suspended).  This means that a participant who does not want to receive rewards can easily avoid them by ignoring one of those rules (the last one is probably the easiest).

File: ledger/eval/eval.go (hunk 1 above)
Comment (jannotti) [reply ledger/eval/eval.go:1222]: Yes, we're going to ban all `FeeSink` transactions.

**Comment by jannotti** [comment]
Closing this just to get out of the way. Intent is to get #5757 in, which has these changes as well.
